extensions/control-plane/provision/provision-oauth2/src/main/java/org/eclipse/edc/connector/provision/oauth2/Oauth2Provisioner.java

```diff
@@ -114,12 +110,38 @@ public CompletableFuture<StatusResult<DeprovisionedResource>> deprovision(Oauth2
     }
 
     @NotNull
-    private String getStringBody(Response response) throws IOException {
-        var body = response.body();
-        if (body != null) {
-            return body.string();
-        } else {
-            return "";
+    private Result<Oauth2CredentialsRequest> createRequest(Oauth2ResourceDefinition rd) {
+        var keySecret = rd.getPrivateKeyName();
+        return keySecret != null ? createPrivateKeyBasedRequest(keySecret, rd) : createSharedSecretRequest(rd);
+    }
+
+    @NotNull
+    private Result<Oauth2CredentialsRequest> createPrivateKeyBasedRequest(String pkSecret, Oauth2ResourceDefinition rd) {
+        return createAssertion(pkSecret, rd)
+                .map(assertion -> PrivateKeyOauth2CredentialsRequest.Builder.newInstance()
+                        .clientAssertion(assertion.getToken())
+                        .url(rd.getTokenUrl())
+                        .grantType(GRANT_CLIENT_CREDENTIALS)
+                        .build());
+    }
+
+    @NotNull
+    private Result<Oauth2CredentialsRequest> createSharedSecretRequest(Oauth2ResourceDefinition rd) {
+        return Result.success(SharedSecretOauth2CredentialsRequest.Builder.newInstance()
+                .url(rd.getTokenUrl())
+                .grantType(GRANT_CLIENT_CREDENTIALS)
+                .clientId(rd.getClientId())
+                .clientSecret(rd.getClientSecret())
+                .build());
+    }
+
+    private Result<TokenRepresentation> createAssertion(String pkSecret, Oauth2ResourceDefinition rd) {
+        var privateKey = privateKeyResolver.resolvePrivateKey(pkSecret, PrivateKey.class);
+        if (privateKey == null) {
+            return Result.failure("Failed to resolve private key with alias: " + pkSecret);
         }
+        var decorator = new Oauth2AssertionDecorator(rd.getTokenUrl(), rd.getClientId(), clock, rd.getValidity());
+        var service = new TokenGenerationServiceImpl(privateKey);
```

Comment:
It can't, as the constructor of `TokenGenerationServiceImpl` takes the private key as argument, while here the private key is "dynamic" in the sense that it is fetched from the `PrivateKeyResolver` based on the data address properties.

Reply:
Oh, you're right, my fault.